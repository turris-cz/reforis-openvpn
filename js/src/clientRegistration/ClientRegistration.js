/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect, useState } from "react";
import PropTypes from "prop-types";

import { useAPIGet, withErrorMessage, withSpinnerOnSending } from "foris";

import API_URLs from "API";
import Clients from "./Clients";
import AddClientForm from "./AddClientForm";

ClientRegistration.propTypes = {
    ws: PropTypes.object.isRequired,
};

export default function ClientRegistration({ ws }) {
    const [authority, getAuthority] = useAPIGet(API_URLs.authority);
    useEffect(() => {
        getAuthority();
    }, [getAuthority]);

    return (
        <>
            <h1>{_("Client Registration")}</h1>
            <RegistrationWithErrorAndSpinner
                apiState={authority.state}
                ws={ws}
                certificateAuthority={authority.data}
            />
        </>
    );
}

const RegistrationWithErrorAndSpinner = withErrorMessage(
    withSpinnerOnSending(Registration)
);

Registration.propTypes = {
    ws: PropTypes.object.isRequired,
    certificateAuthority: PropTypes.object.isRequired,
};

function Registration({ ws, certificateAuthority }) {
    const [generating, setGenerating] = useState(false);

    if (certificateAuthority.status !== "ready") {
        return (
            <p>
                {_(
                    "You need to generate certificate authority in order to register clients."
                )}
            </p>
        );
    }
    return (
        <>
            <p>
                {_(
                    "You need to generate a configuration file for each client that you wish to connect to your OpenVPN server."
                )}
            </p>
            <p>
                {_(
                    "To apply the client configuration you need to download it and put it into the OpenVPN configuration directory or alternatively open it using your OpenVPN client. You might need to restart your client afterwards."
                )}
            </p>
            <AddClientForm
                generating={generating}
                setGenerating={setGenerating}
            />
            <Clients ws={ws} setGenerating={setGenerating} />
        </>
    );
}
