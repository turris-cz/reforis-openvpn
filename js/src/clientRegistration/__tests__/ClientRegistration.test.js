/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import {
    render, wait, getByText, getAllByText, getByLabelText, getAllByRole, queryByText,
    queryByLabelText, fireEvent, act,
} from "customTestRender";
import mockAxios from "jest-mock-axios";

import { WebSockets } from "foris";

import ClientRegistration from "../ClientRegistration";

describe("<ClientRegistration />", () => {
    let webSockets;
    let componentContainer;
    const clients = [
        { id: "A1", name: "First", status: "valid" },
        { id: "A2", name: "Second", status: "generating" },
        { id: "B1", name: "Third", status: "valid" },
        { id: "B2", name: "Last", status: "revoked" },
    ];

    beforeEach(() => {
        webSockets = new WebSockets();
        const { container } = render(<ClientRegistration ws={webSockets} />);
        componentContainer = container;
    });

    it("should display spinner on loading", () => {
        expect(componentContainer).toMatchSnapshot();
    });

    it("should handle API error - authority", async () => {
        mockAxios.mockError({ response: { headers: { "content-type": "application/json" } } });
        await wait(() => getByText(componentContainer, "An error occurred during loading certificate authority details"));
    });

    it("should handle API error - clients", async () => {
        // Response to GET authority
        mockAxios.mockResponse({ data: { status: "ready" } });
        await wait(() => getByText(componentContainer, "Client Registration"));
        mockAxios.mockError({ response: { headers: { "content-type": "application/json" } } });
        await wait(() => getByText(componentContainer, "An error occurred during loading OpenVPN clients"));
    });

    it("should handle invalid CA", async () => {
        mockAxios.mockResponse({ data: { status: "whatever" } });
        await wait(() => getByText(componentContainer, "You need to generate certificate authority in order to register clients."));
    });

    it("should display clients", async () => {
        expect(mockAxios.get).toBeCalledWith("/reforis/openvpn/api/authority", expect.anything());
        // Response to GET authority
        mockAxios.mockResponse({ data: { status: "ready" } });

        await wait(() => getByText(componentContainer, "Client Registration"));

        expect(mockAxios.get).toBeCalledWith("/reforis/openvpn/api/clients", expect.anything());
        // Respose to GET clients
        mockAxios.mockResponse({ data: clients });

        await wait(() => getByText(componentContainer, "First"));
        expect(componentContainer).toMatchSnapshot();
    });

    describe("with mocked clients and authority initial requests", () => {
        function submitClientForm(name) {
            const clientNameInput = getByLabelText(componentContainer, "Client name");
            fireEvent.change(clientNameInput, { target: { value: name } });
            const addButton = getByText(componentContainer, "Add");
            fireEvent.click(addButton);
        }

        function enableOverride() {
            const checkbox = getByLabelText(componentContainer, "Override server address");
            fireEvent.click(checkbox);
        }

        beforeEach(async () => {
            mockAxios.mockResponse({ data: { status: "ready" } });
            await wait(() => getByText(componentContainer, "Client Registration"));
            mockAxios.mockResponse({ data: clients });
            await wait(() => getByText(componentContainer, "Add new client"));
        });

        it("should display address input when checkbox is selected", () => {
            expect(queryByLabelText(componentContainer, "Router's public IPv4 address")).toBeNull();
            enableOverride();
            expect(getByLabelText(componentContainer, "Router's public IPv4 address")).toBeDefined();
        });

        it("should validate server address override", async () => {
            const getFirstDownload = () => getAllByText(componentContainer, "Download configuration")[0];

            enableOverride();
            const addressInput = getByLabelText(componentContainer, "Router's public IPv4 address");
            // Since "a" element is replaced with "button" we have to query for "Download configuration" anew
            expect(getFirstDownload().href).toBe(
                `http://localhost/reforis/openvpn/api/clients/${clients[0].id}?address=`,
            );
            // Test invalid value
            fireEvent.change(addressInput, { target: { value: "999.888.999.888" } });
            expect(getFirstDownload().disabled).toBeTruthy();
            // Test valid value
            const address = "9.8.9.8";
            fireEvent.change(addressInput, { target: { value: address } });
            expect(getFirstDownload().href).toBe(
                `http://localhost/reforis/openvpn/api/clients/${clients[0].id}?address=${address}`,
            );
        });

        it("should add new client to the table", async () => {
            const getSpinnersNumber = () => getAllByRole(componentContainer, "status").length;
            const getDownloadsNumber = () => getAllByText(componentContainer, "Download configuration").length;

            const name = "user";
            submitClientForm(name);
            // Handle request
            expect(mockAxios.post).toBeCalledWith(
                "/reforis/openvpn/api/clients",
                { name },
                expect.anything(),
            );
            mockAxios.mockResponse({ data: { task_id: "1234" } });
            await wait(() => getByText(componentContainer, "Add new client"));

            expect(queryByText(componentContainer, name)).toBeDefined(); // New client not yet visible
            // Generating new client
            const spinners = getSpinnersNumber();
            const downloads = getDownloadsNumber();
            act(() => webSockets.dispatch(
                { module: "openvpn", action: "generate_client", data: { status: "client_generating" } },
            ));
            expect(mockAxios.get).toHaveBeenNthCalledWith(3, "/reforis/openvpn/api/clients", expect.anything());
            const updatedClients = [
                ...clients,
                { id: "C2", name, status: "generating" },
            ];
            mockAxios.mockResponse({ data: updatedClients });
            // New client appears (with spinner)
            await wait(() => expect(getByText(componentContainer, name)).toBeDefined());
            expect(getSpinnersNumber()).toBe(spinners + 1);

            // Client finished generating
            act(() => webSockets.dispatch(
                { module: "openvpn", action: "generate_client", data: { status: "succeeded" } },
            ));
            expect(mockAxios.get).toHaveBeenNthCalledWith(4, "/reforis/openvpn/api/clients", expect.anything());
            updatedClients[4].status = "valid";
            mockAxios.mockResponse({ data: updatedClients });
            await wait(() => expect(getByText(componentContainer, name)).toBeDefined());
            // Spinner disappeared
            expect(getSpinnersNumber()).toBe(spinners);
            // Download is available
            expect(getDownloadsNumber()).toBe(downloads + 1);
        });

        it("should reload clients when one is revoked", async () => {
            const getRevokedNumber = () => getAllByText(componentContainer, "Access revoked").length;

            const revoked = getRevokedNumber();
            // Client is revoked
            act(() => webSockets.dispatch(
                { module: "openvpn", action: "revoke", data: { id: clients[0].id } },
            ));
            expect(mockAxios.get).toHaveBeenNthCalledWith(3, "/reforis/openvpn/api/clients", expect.anything());
            const updatedClients = [...clients];
            updatedClients[0].status = "revoked";
            mockAxios.mockResponse({ data: updatedClients });
            await wait(() => getByText(componentContainer, "First"));
            expect(getRevokedNumber()).toBe(revoked + 1);
        });

        it("should render alerts", async () => {
            const name = "user";
            submitClientForm(name);
            // Handle request
            expect(mockAxios.post).toBeCalledWith(
                "/reforis/openvpn/api/clients",
                { name },
                expect.anything(),
            );
            const errorMessage = "Something went wrong";
            mockAxios.mockError(
                { response: { data: errorMessage, headers: { "content-type": "application/json" } } },
            );
            await wait(() => getByText(componentContainer, "Something went wrong"));
        });
    });
});
