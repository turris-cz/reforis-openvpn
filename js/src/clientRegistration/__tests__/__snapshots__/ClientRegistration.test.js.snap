// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ClientRegistration /> should display clients 1`] = `
<div>
  <h1>
    Client Registration
  </h1>
  <p>
    You need to generate a configuration file for each client that you wish to connect to your OpenVPN server.
  </p>
  <p>
    To apply the client configuration you need to download it and put it into the OpenVPN configuration directory or alternatively open it using your OpenVPN client. You might need to restart your client afterwards.
  </p>
  <div
    class="card p-4 col-sm-12 col-lg-12 p-0 mb-4"
  >
    <h2>
      Add new client
    </h2>
    <form>
      <div
        class="form-group"
      >
        <label
          for="1"
        >
          Client name
        </label>
        <div
          class="input-group"
        >
          <input
            class="form-control  is-invalid"
            id="1"
            type="text"
            value=""
          />
        </div>
        <div
          class="invalid-feedback"
        >
          Name cannot be empty
        </div>
        <small
          class="form-text text-muted"
        >
          Shorter than 64 characters. Only alphanumeric characters, dots, dashes and underscores.
        </small>
      </div>
      <div
        class="text-right"
      >
        <button
          class="btn btn-primary  col-sm-12 col-md-3 col-lg-2"
          disabled=""
          type="submit"
        >
          Add
        </button>
      </div>
    </form>
    <h2>
      Client configurations
    </h2>
    <p>
      Be sure to check if server's IP address provided in configuration file actually matches the public IP address of your router. You can set this address manually if the autodetection fails. This change is 
      <strong>
        not
      </strong>
       stored anywhere and is applicable only to the configuration being currently downloaded.
    </p>
    <div>
      <div
        class="form-group"
      >
        <div
          class="custom-control custom-checkbox "
        >
          <input
            class="custom-control-input"
            id="2"
            type="checkbox"
            value="false"
          />
          <label
            class="custom-control-label"
            for="2"
          >
            Override server address
          </label>
        </div>
      </div>
      <div
        class="table-responsive"
      >
        <table
          class="table table-hover"
        >
          <thead
            class="thead-light"
          >
            <tr>
              <th
                scope="col"
              >
                Client name
              </th>
              <th
                aria-label="Actions"
                scope="col"
              />
            </tr>
          </thead>
          <tbody>
            <tr>
              <td
                class="align-middle"
              >
                First
              </td>
              <td
                class="text-right"
              >
                <div
                  class="btn-group mb-0"
                  role="group"
                >
                  <a
                    class="btn btn-primary btn-sm"
                    download=""
                    href="/reforis/openvpn/api/clients/A1?address="
                  >
                    Download
                  </a>
                  <button
                    class="btn btn-danger btn-sm"
                    type="button"
                  >
                    Revoke
                  </button>
                </div>
              </td>
            </tr>
            <tr>
              <td
                class="align-middle"
              >
                Second
              </td>
              <td
                class="text-right"
              >
                <div
                  class="spinner-border spinner-border-sm"
                  role="status"
                >
                  <span
                    class="sr-only"
                  />
                </div>
              </td>
            </tr>
            <tr>
              <td
                class="align-middle"
              >
                Third
              </td>
              <td
                class="text-right"
              >
                <div
                  class="btn-group mb-0"
                  role="group"
                >
                  <a
                    class="btn btn-primary btn-sm"
                    download=""
                    href="/reforis/openvpn/api/clients/B1?address="
                  >
                    Download
                  </a>
                  <button
                    class="btn btn-danger btn-sm"
                    type="button"
                  >
                    Revoke
                  </button>
                </div>
              </td>
            </tr>
            <tr>
              <td
                class="align-middle"
              >
                Last
              </td>
              <td
                class="text-right"
              >
                <span
                  class="text-muted"
                >
                  Access revoked
                </span>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;

exports[`<ClientRegistration /> should display spinner on loading 1`] = `
<div>
  <h1>
    Client Registration
  </h1>
  <div
    class="spinner-wrapper my-3 text-center"
  >
    <div
      class="spinner-border"
      role="status"
    >
      <span
        class="sr-only"
      />
    </div>
  </div>
</div>
`;
